Introduction to Temporal update (#728)
temporalio/documentation: docs/content/why-temporal.md

- A traditional function execution typically has a limited lifespan, often measured in minutes.

With Temporal, a Workflow Execution is resumable.

- A Workflow Execution is fully resumable after a failure.
- Temporal imposes no deadlines on Workflow Executions.

**State**

With a traditional system, stoppage or failure means that all execution state is lost.
Your application (or a supporting component) must monitor the service's response to initiate a retry of the service execution.
A retry starts from its _initial_ state.

With Temporal, computation resumes from its _latest_ state. All progress is retained.

**Communication**

With a traditional system, you can't communicate with a function execution.

With Temporal, Signals and Queries enable data to be sent to or extracted from a Workflow Execution.

**Scope**

With a traditional system, a service function execution can at best represent a business process.
Typically, it represents only a part of a business process.

A Temporal Workflow Execution can represent a business process or an entire business object.

> ### Example subscription use-case

Reviewer: Find-replace use-case -> "use case"? I believe the latter is much more common. https://en.wikipedia.org/wiki/Use_case

> Let's look at a subscription-based use-case to compare the difference between a Temporal Application and other traditional modern approaches.

Reviewer: "traditional modern" again

The basic business steps are as follows:

1. A customer signs up for a service that has a trial period.

> 2. After the trial period, if the customer has not cancelled, they should be charged once a month, for up to X months.

Reviewer: ```suggestion
2. After the trial period, if the customer has not canceled, they should be charged once a month.
```

3. The customer has to be notified via email about the charges and should be able to cancel the subscription at any time.

This business logic is not very complicated and can be expressed in a few dozen lines of code.

> In addition to that, any practical implementation has to ensure that the business process is fault-tolerant and scalable.

Reviewer: ```suggestion
Any practical implementation also has to ensure that the business process is fault-tolerant and scalable.
```

Avoids slightly awk "that" antecedent

2. Failed calls likely need to be retried for a long time, and these calls need to be throttled to not overload external resources.
3. There needs to be logic to handle corrupted customer records to avoid blocking the whole process.
4. Additionally, databases have performance and scalability limitations (eventually requiring sharding) and are not efficient for scenarios that require constant polling.

**Queue system design approach**

The next commonly employed approach is to use a timer service and queues.
Updates are pushed to a queue while a service consumes them one at a time, updating a database, and possibly pushing more messages into other downstream queues.
A timer service can be used to schedule queue polling or database actions.

While this approach has shown to scale a bit better, the programming model can become very complex and error-prone, as there are usually no transactional updates between a queuing system, a timer service, and a database.

**Temporal design approach**

The Temporal Platform approach aims to encapsulate and implement the entire business logic in a simple function or object method.

> Thanks to the Temporal Platform, the function/method is stateful, and the implementer doesn't need to employ any additional systems to ensure durability and fault tolerance.

Reviewer: ```suggestion
Thanks to the Temporal Platform, the function/method is durably stateful, and the implementer doesn't need to employ any additional systems to ensure consistency and fault tolerance.
```

My reaction to "method is stateful" is like "of course methods have state/variables". Want to get across that each step/state change is persisted.

Here are example Workflow Definitions that implement the subscription management use case in Java, Go, and PHP:

> <Tabs
> defaultValue="java"
> values={[
> {label: 'Go', value: 'go'},
> {label: 'Java', value: 'java'},
> {label: 'PHP', value: 'php'},
> ]
> }>
> 
> <TabItem value="go">
> 
> <!--SNIPSTART subscription-go-workflow-definition-->
> <!--SNIPEND-->
> 
> </TabItem>
> <TabItem value="java">
> 
> <!--SNIPSTART subscription-java-workflow-definition-implementation-->
> <!--SNIPEND-->
> 
> </TabItem>
> <TabItem value="php">
> 
> <!--SNIPSTART subscription-php-workflow-definition-implementation-->
> <!--SNIPEND-->
> 
> </TabItem>
> </Tabs>

Reviewer 2: i have ported the subscription example to Typescript (https://github.com/temporalio/subscription-workflow-project-template-typescript) however i cannot find where the snipsync snippets go for subscription-ts-workflow-definition-implementation. could you help?

```suggestion
<Tabs
defaultValue="java"
values={[
{label: 'Go', value: 'go'},
{label: 'Java', value: 'java'},
{label: 'TypeScript', value: 'ts'},
{label: 'PHP', value: 'php'},
]
}>

<TabItem value="go">

<!--SNIPSTART subscription-go-workflow-definition-->
<!--SNIPEND-->

</TabItem>
<TabItem value="java">

<!--SNIPSTART subscription-java-workflow-definition-implementation-->
<!--SNIPEND-->

</TabItem>
<TabItem value="ts">

<!--SNIPSTART subscription-ts-workflow-definition-implementation-->
<!--SNIPEND-->

</TabItem>
<TabItem value="php">

<!--SNIPSTART subscription-php-workflow-definition-implementation-->
<!--SNIPEND-->

</TabItem>
</Tabs>
```

Author: Got it working.

Again, it is important to note that this is working application code that directly implements the business logic.
If any of the operations (aka [Activities](/docs/concepts/activities)) take a long time, the code is not going to change.

It is completely okay to be blocked on `chargeCustomerForBillingPeriod` for a day or more if the downstream processing service is down or not responding.
In the same way, it is a completely normal operation to sleep for 30 days directly inside the Workflow code.

> This is possible because infrastructure failures are not going to affect the Workflow state including threads, blocking calls, and any local or Workflow variables.

Reviewer: > local or Workflow variables.

Reader might not know what this means—maybe simplify to "any variables"?

---
id: why-temporal
title: What are the differences between Temporal and a traditional system?
description: This page compares a function execution in a traditional system and a Workflow Execution in Temporal.
tags:
  - explanation
---

import CenteredImage from "../components/CenteredImage.js"
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

<!-- prettier-ignore -->
import * as WhatIsAWorkflowExecution from '../content/what-is-a-workflow-execution.md'
import * as WhatIsASignal from '../content/what-is-a-signal.md'
import * as WhatIsAQuery from '../content/what-is-a-query.md'

One of the aspects of the Temporal System is that it abstracts the complexity of a distributed system.
Distributed systems exist to scale computation across multiple machines as the potential load of a system changes.

> In theory a distributed system facilitates a reliable and highly performant application.

Reviewer 3: ```suggestion
In theory, a distributed system facilitates a reliable and highly performant application.
```

> However there are a set of failures that can occur in a distributed application that can make things very complicated, especially at a large scale.

Reviewer 3: ```suggestion
However, a set of failures can occur in a distributed application that can make things very complicated, especially at a large scale.
```

Author: ```suggestion
However, a set of failures can occur in a distributed application that can make things very complicated, especially at a large scale.
```
just seeing if it will let me commit my suggestion - as it wont let me commit yours.

---
id: why-temporal
title: What are the differences between Temporal and a traditional system?
description: This page compares a function execution in a traditional system and a Workflow Execution in Temporal.
tags:
  - explanation
---

import CenteredImage from "../components/CenteredImage.js"
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

<!-- prettier-ignore -->
import * as WhatIsAWorkflowExecution from '../content/what-is-a-workflow-execution.md'
import * as WhatIsASignal from '../content/what-is-a-signal.md'
import * as WhatIsAQuery from '../content/what-is-a-query.md'

One of the aspects of the Temporal System is that it abstracts the complexity of a distributed system.
Distributed systems exist to scale computation across multiple machines as the potential load of a system changes.
In theory a distributed system facilitates a reliable and highly performant application.

However there are a set of failures that can occur in a distributed application that can make things very complicated, especially at a large scale.

<CenteredImage
imagePath="/diagrams/basic-distributed-system.svg"
imageSize="75"
title="Distributed application failures"
/>

> How will a downstream part of the application know if there was a failure before, failure during, failure between, or a failure after the changes to the state?

Reviewer 3: ```suggestion
How will a downstream part of the application know if there was a failure before, failure during, failure between, or failure after the changes to the state?
```
Alternatively, add "a" to the middle two instances of "failure".

> How is any downstream part of the application supposed to know if there was a failure before, failure during, failure between, or a failure after the changes to the state?

Author: ```suggestion
How will a downstream part of the application know if there was a failure before, failure during, failure between, or a failure after the changes to the state?
```

How will the inconsistent state be reconciled if there was a failure between changes?

In traditional systems, a large investment is often made to maintain the health of each individual component, visualize the health of the overall system, define timeout constraints for computations, orchestrate retries for computations that fail, and maintain a consistent state.

These systems are often a mixture of stateless services, databases, cron jobs, and queues.
And as these systems scale, responding to multiple asynchronous events, communicating with unreliable external resources, or tracking the state of something very complex becomes more and more challenging.

> Temporal reconfigures the use of services, databases, cron jobs, and queues into the Temporal Platform.

Reviewer 3: Instead of "reconfigures", perhaps use "integrates"?

Author: Not sure integrates works here. Maybe "restructures" or "re-envisions" - and add SDKs and host processes to the list to complete the picture.
```suggestion
Temporal restructures the use of services, databases, cron jobs, queues, host processes, and SDKs, into the Temporal Platform.
```

> The Temporal Platform addresses these failures head on and right out the box.

Reviewer 3: We haven't discussed failures for three paragraphs, so the reference to "these failures" is ambiguous. Perhaps delete this sentence and extend the preceding sentence with "...and addresses failures head on."

(Side note: Our style guides inveigh against using "out [of] the box.")

> In a traditional system the service exists to spawn function executions.

Reviewer 3: ```suggestion
In a traditional system, the service exists to spawn function executions.
```

Typically, it represents only a part of a business process.

A Temporal Workflow Execution can represent a business process or an entire business object.

### Example subscription use case

Let's look at a subscription-based use case to compare the difference between a Temporal Application and other traditional approaches.

The basic business steps are as follows:

1. A customer signs up for a service that has a trial period.
2. After the trial period, if the customer has not canceled, they should be charged once a month.
3. The customer has to be notified via email about the charges and should be able to cancel the subscription at any time.

This business logic is not very complicated and can be expressed in a few dozen lines of code.
Any practical implementation also has to ensure that the business process is fault-tolerant and scalable.

**Database-centric design approach**

The first approach might be to center everything around a database where an application process would periodically scan the database tables for customers in specific states, execute necessary actions, and update the database to reflect changes.

However, there are various drawbacks.

> 1. The most obvious one is that the application state machine of the customer's state quickly becomes extremely complicated.

Reviewer 3: This ordered list should likely be unordered.

Here are example Workflow Definitions that implement the subscription management use case in Java, Go, and PHP:

<Tabs
defaultValue="java"
values={[
{label: 'Go', value: 'go'},
{label: 'Java', value: 'java'},
{label: 'Typescript', value: 'ts'},
{label: 'PHP', value: 'php'},
]
}>

<TabItem value="go">

<!--SNIPSTART subscription-go-workflow-definition-->
<!--SNIPEND-->

</TabItem>
<TabItem value="java">

<!--SNIPSTART subscription-java-workflow-definition-implementation-->
<!--SNIPEND-->

</TabItem>
<TabItem value="ts">

<!--SNIPSTART subscription-ts-workflow-definition-->
<!--SNIPEND-->

</TabItem>
<TabItem value="php">

<!--SNIPSTART subscription-php-workflow-definition-implementation-->
<!--SNIPEND-->

</TabItem>
</Tabs>

Again, it is important to note that this is working application code that directly implements the business logic.
If any of the operations take a long time, the code is not going to change.

It is completely okay to be blocked on `chargeCustomerForBillingPeriod` for a day or more if the downstream processing service is down or not responding.
In the same way, it is a completely normal operation to sleep for 30 days directly inside the Workflow code.

> This is possible because infrastructure failures are not going to affect the Workflow state – including threads, blocking calls, and any variables.

Reviewer 3: ```suggestion
This is possible because infrastructure failures won't affect the Workflow state—including threads, blocking calls, and any variables.
```
